From e043fcbb7e390e8f53da7d0b5565f8e02e07634f Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Sat, 27 Sep 2014 21:44:36 +0800
Subject: [PATCH 0126/5242] MLK-9638-5 usb: chipidea: otg: test: timer for B
 host hands host role back

commit  8ae012b8b59b8189cd3ac568e87d6bcd590d1ba3 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch implements a timer for B host to start host role switch back to A
device.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 139bcc46841f73199bdc3ca308e46a5af3dd2fc1)
(cherry picked from commit d1c81fc3fa2f8cc141d688b6f627ffa6ca4a7204)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/otg_fsm.c |   16 ++++++++++++++++
 drivers/usb/chipidea/otg_fsm.h |    5 +++++
 2 files changed, 21 insertions(+)

diff --git a/drivers/usb/chipidea/otg_fsm.c b/drivers/usb/chipidea/otg_fsm.c
index 0251362..fa85e12 100644
--- a/drivers/usb/chipidea/otg_fsm.c
+++ b/drivers/usb/chipidea/otg_fsm.c
@@ -217,6 +217,7 @@
 	TA_DP_END,
 	TA_TST_MAINT,
 	TB_SRP_REQD,
+	TB_TST_SUSP,
 	0,
 };
 
@@ -401,6 +402,20 @@ static int b_srp_reqd_tmout(struct ci_hdrc *ci)
 }
 
 /*
+ * otg_hnp_reqd feature
+ * After B(UUT) switch to host, B should hand host role back
+ * to A(PET) within TB_TST_SUSP after setting configuration.
+ */
+static int b_tst_susp_tmout(struct ci_hdrc *ci)
+{
+	if (ci->fsm.otg->state == OTG_STATE_B_HOST) {
+		ci->fsm.b_bus_req = 0;
+		return 0;
+	}
+	return 1;
+}
+
+/*
  * Keep this list in the same order as timers indexed
  * by enum otg_fsm_timer in include/linux/usb/otg-fsm.h
  */
@@ -420,6 +435,7 @@ static int (*otg_timer_handlers[])(struct ci_hdrc *) = {
 	a_dp_end_tmout,		/* A_DP_END */
 	a_tst_maint_tmout,	/* A_TST_MAINT */
 	b_srp_reqd_tmout,	/* B_SRP_REQD */
+	b_tst_susp_tmout,	/* B_TST_SUSP */
 	NULL,			/* HNP_POLLING */
 };
 
diff --git a/drivers/usb/chipidea/otg_fsm.h b/drivers/usb/chipidea/otg_fsm.h
index 867c40c..5d9c0e6 100644
--- a/drivers/usb/chipidea/otg_fsm.h
+++ b/drivers/usb/chipidea/otg_fsm.h
@@ -69,6 +69,11 @@
 					 * pulse after A(PET) turn off v-bus
 					 */
 
+#define TB_TST_SUSP          (20)	/* B-dev hand host role back to A-dev
+					 * via suspend bus after set config.
+					 * max: 100ms
+					 */
+
 #if IS_ENABLED(CONFIG_USB_OTG_FSM)
 
 int ci_hdrc_otg_fsm_init(struct ci_hdrc *ci);
-- 
1.7.9.5

